src/git/smart.mli

```diff
@@ -0,0 +1,616 @@
+
+  type request_command =
+    [ `UploadPack (** When the client wants to fetch/clone. *)
+    | `ReceivePack (** When the client wants to push. *)
+    | `UploadArchive (** When the client wants an archive of the remote git repository. *)
+    ] (** The Git command. *)
+
+  type git_proto_request =
+    { pathname        : string
+    ; host            : (string * int option) option
+    ; request_command : request_command }
+  (** The Git transport starts off by sending the command and the repository
+      on the wire using the {i packet-line} format, followed by a NUL byte and a
+      hostname parameter, terminated by a NUL byte.deep
+
+      This type represents this information. *)
+
+  type ('a, 'b) either =
+    | L of 'a
+    | R of 'b
+  and update_request =
+    { shallow      : Hash.t list
+    ; requests     : (command * command list, push_certificate) either
+    ; capabilities : capability list }
+  (** Once the client knows what the references the server is at, it can send
+      a list of reference update requests. For each reference on the server that
+      it wants to update, it sends a line listing the obj-id currently on the
+      server, the obj-id the client would like to update it to and the name of the
+      reference.
+
+      This type represents this information. *)
+  and command =
+    | Create of Hash.t * string (** When the client wants to create a new reference. *)
+    | Delete of Hash.t * string (** When the client wants to delete an existing reference in the server side. *)
+    | Update of Hash.t * Hash.t * string (** When the client wants to update an existing reference in the server-side. *)
+  and push_certificate =
+    { pusher   : string
+    ; pushee   : string
+    ; nonce    : string
+    ; options  : string list
+    ; commands : command list
+    ; gpg      : string list }
+
+  type action =
+    [ `GitProtoRequest of git_proto_request
+    | `UploadRequest of upload_request
+    | `UpdateRequest of update_request
+    | `Has of Hash.t list
+    | `Done
+    | `Flush
+    | `Shallow of Hash.t list
+    | `PACK of int ]
+  (** The type action to describe what is expected to encode/send. *)
+
+  val encode : encoder -> action -> unit state
+  (** [encode encoder action] encodes to an output represented by [encoder] in
+      the way of the [action] and returns an unit value. *)
+
+  val encoder : unit -> encoder
+  (** [encoder ()] makes a new decoder. *)
+end
+
+module Encoder
+    (Digest : Ihash.IDIGEST with type t = Bytes.t)
+  : ENCODER with type Hash.t = Digest.t
+             and module Digest = Digest
+(** The {i functor} to make the Encoder by a specific hash implementation. We
+    constraint the {!IDIGEST} module to generate a {!Bytes.t}. *)
+
+module type CLIENT =
+sig
+  module Digest
+    : Ihash.IDIGEST with type t = Bytes.t
+  (** The [Digest] module used to make the module. *)
+
+  module Decoder
+    : DECODER with type Hash.t = Digest.t and module Digest = Digest
```

Comment:
The Smart decoder (and the encoder too) needs to know the length of the current hash implementation and the `Digest` module notify this information with `Digest.length`. I can reduce the signature need by these functors (like: `sig val length : int end`) if you want.